src/common/blob/util.js

```diff
@@ -98,12 +99,126 @@ define(['q', './BlobMetadata'], function (Q, BlobMetadata) {
+                    var softLinkMetadataHash = metadata.content[softLinkName].content;
+                    logger.debug('Complex object, softLinkMetadataHash:', softLinkMetadataHash);
+                    return Q.ninvoke(blobClient, 'getMetadata', softLinkMetadataHash)
+                        .then(function (softLinkMetadata) {
+                            return _gatherFilesFromMetadataHashRec(logger, blobClient,
+                                softLinkMetadata, softLinkMetadataHash, artifact);
+                        });
+                }))
+                .then(function () {
+                    // Finally add the .metadata for the complex object.
+                    return Q.ninvoke(artifact, 'addFile', filenameMetadata, JSON.stringify(metadata));
+                })
+                .then(function () {
+                    deferred.resolve();
+                })
+                .catch(function (err) {
+                    deferred.reject(new Error(err));
+                });
+        } else {
+            deferred.reject(new Error('Unsupported content type', metadata.contentType));
+        }
+
+        return deferred.promise;
+    }
+
+    /**
+     *
+     * @param {GmeLogger} logger
+     * @param {BlobClient} blobClient
+     * @param {object} jsonExport
+     * @param {boolean} addAssets
+     * @param callback
+     * @returns {*}
+     */
+    function buildProjectPackage(logger, blobClient, jsonExport, addAssets, callback) {
+        console.log('exporting.', addAssets);
+        var artie = blobClient.createArtifact(jsonExport.projectId +
+                '_' + (jsonExport.branchName || jsonExport.commitHash)),
+            assets = jsonExport.hashes.assets || [];
+
+        artie.descriptor.name = jsonExport.projectId +
+            '_' + (jsonExport.branchName || jsonExport.commitHash) + '.webgmeX';
+        console.log('exporting:', assets);
+
+        if (!addAssets) {
+            assets = [];
+        }
+
+        Q.allSettled(assets.map(function (assetHash) {
+                return Q.ninvoke(blobClient, 'getMetadata', assetHash)
+                    .then(function (metadata) {
+                        return _gatherFilesFromMetadataHashRec(logger, blobClient, metadata, assetHash, artie);
+                    });
+            }))
+            .then(function (result) {
+                var error,
+                    i;
+
+                for (i = 0; i < result.length; i += 1) {
+                    if (result[i].state === 'rejected') {
+                        error = result[i].reason instanceof Error ? result[i].reason : new Error(result[i].reason);
+                        logger.debug('Gathering returned with error', assets[i], error);
+                        if (error.message.indexOf('Another content with the same name was already added.') === -1) {
+                            //some real error
+                            logger.error('Error for gathering asset failed:', assets[i], error);
```

Comment:
No need to log the error (since there's no extra info I would let the caller log it).